clients/src/main/java/org/apache/kafka/clients/producer/internals/TransactionManager.java

```diff
@@ -476,29 +519,67 @@ private void setProducerIdAndEpoch(ProducerIdAndEpoch producerIdAndEpoch) {
      * would not have any way of knowing this happened. So for the transactional producer, it's best to return the
      * produce error to the user and let them abort the transaction and close the producer explicitly.
      */
-    synchronized void resetIdempotentProducerId() {
+    private synchronized void resetIdempotentProducerId() {
```

Comment:
We don't necessarily have to fix this here, but on going through this code again, I found this method a bit confusing. We call it from `bumpIdempotentProducerEpoch` in order to transition to UNINITIALIZED so that the logic in `bumpIdempotentProducerEpochIfNeeded` kicks in and we send the InitProducerId request. Seems like it would be more direct to send InitProducerId from here. 

Note also, that the comment above seems a little dated. Particularly the comment about bumping the epoch.

Reply:
Fixed the comment. The only issue I see with calling InitProducerId directly here is that we still need to call it outside the epoch bump check the first time through, since this is where the idempotent producer ID gets initialized the first time.